File: 0518-coin-change-ii/0518-coin-change-ii.cpp
class Solution {
public:
    int helper(int index,int amount,vector<int>&coins,vector<vector<int>>&dp){
        if(index==0){
            if(amount%coins[0]==0){
                return 1;
            }
            return 0;
        }
        if(dp[index][amount]!=-1)return dp[index][amount];
        int not_take=helper(index-1,amount,coins,dp);
        int take=0;
        if(amount>=coins[index]){
            take=helper(index,amount-coins[index],coins,dp);
        }
        return dp[index][amount]=take+not_take;
    }
    int change(int amount, vector<int>& coins) {
        int n=coins.size();
        vector<vector<int>>dp(n,vector<int>(amount+1,-1));
        return helper(n-1,amount,coins,dp);
    }
};

File: 0518-coin-change-ii/0518-coin-change-ii.java
class Solution {
    public int helper(int index,int[]coins,int[][]dp,int amount){
        if(amount==0)return 1;
        if(index==coins.length-1){
            if(amount%coins[index]==0)return 1;
            else return 0;
        }
        if(dp[index][amount]!=-1)return dp[index][amount];
        int not_take=helper(index+1,coins,dp,amount);
        int take=0;
        if(coins[index]<=amount)take=helper(index,coins,dp,amount-coins[index]);
        return dp[index][amount]=take+not_take;
    }
    public int change(int amount, int[] coins) {
        int[][]dp=new int[coins.length][amount+1];
        for (int i = 0; i < coins.length; i++) {
            Arrays.fill(dp[i], -1);
        }
        return helper(0,coins,dp,amount);
    }
}